compiler/rustc_span/src/lib.rs

```diff
@@ -1957,22 +1983,35 @@ where
     fn hash_stable(&self, ctx: &mut CTX, hasher: &mut StableHasher) {
         const TAG_VALID_SPAN: u8 = 0;
         const TAG_INVALID_SPAN: u8 = 1;
+        const TAG_RELATIVE_SPAN: u8 = 2;
 
         if !ctx.hash_spans() {
             return;
         }
 
-        self.ctxt().hash_stable(ctx, hasher);
+        let span = self.data_untracked();
+        span.ctxt.hash_stable(ctx, hasher);
+        span.parent.hash_stable(ctx, hasher);
 
-        if self.is_dummy() {
+        if span.is_dummy() {
             Hash::hash(&TAG_INVALID_SPAN, hasher);
             return;
         }
 
+        if let Some(parent) = span.parent {
+            let def_span = ctx.def_span(parent).data_untracked();
+            if def_span.contains(span) {
```

Comment:
This `contains` is just a heuristic right?
We are assuming that if the span is contained in the parent's span, then it will usually move together with it.
And if it's not contained, then maybe some macros are involved, and it will likely not move together with the parent item, so it's better to not enable the incremental optimization.

Reply:
Yes, this is my reasoning.